pkg/controller/master-controller-manager/rbac/sync_resource_test.go

```diff
@@ -1876,3 +1882,303 @@ func TestEnsureProjectClusterRBACRoleForNamedResource(t *testing.T) {
+						},
+					},
+					RoleRef: rbacv1.RoleRef{
+						APIGroup: rbacv1.GroupName,
+						Kind:     "Role",
+						Name:     "kubermatic:constraint:owners",
+					},
+				},
+			},
+			existingRoleBindings: []*rbacv1.RoleBinding{
+				{
+					ObjectMeta: metav1.ObjectMeta{
+						Name:            "kubermatic:constraint:owners",
+						Namespace:       "cluster-clusterid",
+						ResourceVersion: "1",
+					},
+					Subjects: []rbacv1.Subject{
+						{
+							APIGroup: rbacv1.GroupName,
+							Kind:     "Group",
+							Name:     "editors-my-first-project",
+						},
+					},
+					RoleRef: rbacv1.RoleRef{
+						APIGroup: rbacv1.GroupName,
+						Kind:     "Role",
+						Name:     "kubermatic:constraint:owners",
+					},
+				},
+			},
+		},
+	}
+
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			// setup the test scenario
+			ctx := context.Background()
+
+			objs := []ctrlruntimeclient.Object{test.dependantToSync}
+			for _, existingRole := range test.existingRoles {
+				objs = append(objs, existingRole)
+			}
+
+			for _, existingRoleBinding := range test.existingRoleBindings {
+				objs = append(objs, existingRoleBinding)
+			}
+
+			fakeMasterClusterClient := fakectrlruntimeclient.NewClientBuilder().WithObjects(objs...).Build()
+			// act
+			target := resourcesController{
+				client:     fakeMasterClusterClient,
+				restMapper: getFakeRestMapper(t),
+				objectType: test.dependantToSync.DeepCopyObject().(ctrlruntimeclient.Object),
+			}
+			objmeta, err := meta.Accessor(test.dependantToSync)
+			assert.NoError(t, err)
+			_, err = target.Reconcile(ctx, reconcile.Request{NamespacedName: types.NamespacedName{
+				Namespace: objmeta.GetNamespace(),
+				Name:      objmeta.GetName(),
+			}})
+
+			// validate
+			if !test.expectError {
+				assert.NoError(t, err)
+			}
+			if test.expectError {
+				assert.Error(t, err)
+				return
+			}
+
+			{
+				var roles rbacv1.RoleList
+				err = fakeMasterClusterClient.List(context.Background(), &roles)
+				assert.NoError(t, err)
+
+			expectedRolesLoop:
+				for _, expectedRole := range test.expectedRoles {
+					// double-iterating over both slices might not be the most efficient way
+					// but it spares the trouble of converting pointers to values
+					// and then sorting everything for the comparison.
+
+					for _, existingRole := range roles.Items {
+						if strings.EqualFold(expectedRole.Name, existingRole.Name) && reflect.DeepEqual(expectedRole.Rules, existingRole.Rules) {
+							continue expectedRolesLoop
+
+						}
+					}
+					t.Fatalf("expected Role %q not found in cluster", expectedRole.Name)
+				}
+			}
+
+			{
+				var roleBindings rbacv1.RoleBindingList
+				err = fakeMasterClusterClient.List(context.Background(), &roleBindings)
+				assert.NoError(t, err)
+
+			expectedRoleBindingsLoop:
+				for _, expectedRoleBinding := range test.expectedRoleBindings {
+					// double-iterating over both slices might not be the most efficient way
```

Comment:
The nested iteration is not really a concern here, but if you want to avoid it and cover the creation of unwanted resources too: you could use put both actual and wanted roles in a map keyed by name, and use `deep.Equal` to compare them. This gives you a diff that could be useful to debug test failures. e.g.

https://github.com/kubermatic/kubermatic/blob/9f2af8e327595efbe90b96fd8f44bd51dcafa7ff/pkg/controller/nodeport-proxy/envoymanager/controller_test.go#L372-L374

Not a blocker, take it as a suggestion ;-)

Reply:
thats a good suggestion. Let me check the diff. 

The nested iteration is a product of the other tests in the file. I wanted to keep it similar.

Reply:
Ok, i did something similar to what you suggest, but instead of comparing whole result, i just compare those we are interested about in the test(constraint ones).